src/Program2.cs

```diff
@@ -0,0 +1,526 @@
+//                         // call to WaitForNext returns false
+//    }
+
+//    Task Slow() => throw null;
+//}
+
+class Program
+{
+    static async Task Main()
+    {
+        IAsyncEnumerable<int> test = AsyncIterator();
+        IAsyncEnumerator<int> enumerable = test.GetAsyncEnumerator();
+
+        //foreach await (var value in enumerable)
+        //{
+        //    Console.WriteLine(value)
+        //}
+        while (await enumerable.WaitForNextAsync())
+        {
+            while (true)
+            {
+                int value = enumerable.TryGetNext(out bool success);
+                if (!success) goto outer_loop_continue;
+
+                Console.WriteLine(value);
+            }
+            outer_loop_continue:;
+        }
+    }
+
+    static IAsyncEnumerable<int> AsyncIterator()
+    {
+        var stateMachine = new Unprounouncable();
+        stateMachine.State = StateMachineStates.NotStartedStateMachine;
+        stateMachine.Builder = AsyncTaskMethodBuilder.Create();
+        stateMachine._valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(stateMachine);
+        stateMachine._promiseIsActive = true;
+
+        // note: we don't start the machine.
+        return stateMachine;
+    }
+
+    // TODO what happens to exceptions? (should they bubble into value promise, or anywhere else?)
+    private sealed class Unprounouncable :
+        IAsyncStateMachine,
+        IAsyncEnumerable<int>,
+        IAsyncEnumerator<int>,
+        IValueTaskSource<bool>, // used as the backing store behind the ValueTask<bool> returned from each MoveNextAsync
+        IStrongBox<ManualResetValueTaskSourceLogic<bool>> // exposes its ValueTaskSource logic implementation
+    {
+        // If you need to check for states other than NotStarted or Finished, then you must first ensure the promise is not set (ie. _promiseIsActive is false). Otherwise, the machine may be running on another thread.
+        public int State;
+
+        // awaiter and builder are used for 'await' in the method body
+        public TaskAwaiter Awaiter; // thin wrapper around a Task (that we'll be waiting on)
+        public AsyncTaskMethodBuilder Builder; // used for getting a callback to MoveNext when async code completes
+
+        private int _current;
+
+        #region promise
+        // All of the logic for managing the IValueTaskSource implementation
+        public ManualResetValueTaskSourceLogic<bool> _valueOrEndPromise; // promise for a value or end (true means value found, false means finished state)
+        public bool _promiseIsActive = false; // this is spiritually equivalent to saying the promise is set versus null
+
+        ref ManualResetValueTaskSourceLogic<bool> IStrongBox<ManualResetValueTaskSourceLogic<bool>>.Value
+            => ref _valueOrEndPromise;
+        ValueTaskSourceStatus IValueTaskSource<bool>.GetStatus(short token)
+            => _valueOrEndPromise.GetStatus(token);
```

Comment:
@stephentoub Is there a reason `ManualResetValueTaskSourceLogic` doesn't implement `IValueTaskSource<bool>`?
Would that occur a boxing when we trying to call its methods?

Reply:
What benefit would that provide?  While it could do so, I couldn't think of a scenario where that would be beneficial, e.g. you wouldn't want to store it in a field typed as `IValueTaskSource<bool>` as then it would get boxed.

Reply:
The reason I'm asking is this would reduce the number of well-known members for the compiler. The compiler would know about `IValueTaskSource<bool>.GetResult` but not `ManualResetValueTaskSourceLogic.GetResult` for instance.

@gafter You mentioned it is possible to call an interface method on a struct without boxing. Do you have any pointer of where we already do that, by any chance?

Reply:
It's possible to call it, but where would you store it?  The field in your class would need to be typed as `ManualResetValueTaskSourceLogic` (or as a generic which you then construct with it); if you typed the field as `IValueTaskSource<bool>`, there's no way around boxing it.

Reply:
It would be stored as a struct (a `ManualResetValueTaskSourceLogic`). Neal mentioned there is an IL instruction to invoke an interface method on a struct without boxing. I'll try to confirm.

Reply:
> It would be stored as a struct (a ManualResetValueTaskSourceLogic)

At which point doesn't the compiler already need to know about it as a well-known type?

> Neal mentioned there is an IL instruction to invoke an interface method on a struct without boxing

Yes, a contrained call.  The compiler uses this, for example, with `using` for the Dispose call.